instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshOnDemandTest.kt

```diff
@@ -0,0 +1,286 @@
+import com.mapbox.navigation.instrumentation_tests.utils.DynamicResponseModifier
+import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.getSuccessfulResultOrThrowException
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.requestRoutes
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.routesUpdates
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.sdkTest
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.setNavigationRoutesAndWaitForUpdate
+import com.mapbox.navigation.instrumentation_tests.utils.http.FailByRequestMockRequestHandler
+import com.mapbox.navigation.instrumentation_tests.utils.http.MockDirectionsRefreshHandler
+import com.mapbox.navigation.instrumentation_tests.utils.http.MockDirectionsRequestHandler
+import com.mapbox.navigation.instrumentation_tests.utils.http.MockRoutingTileEndpointErrorRequestHandler
+import com.mapbox.navigation.instrumentation_tests.utils.location.MockLocationReplayerRule
+import com.mapbox.navigation.instrumentation_tests.utils.readRawFileText
+import com.mapbox.navigation.testing.ui.BaseTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.take
+import kotlinx.coroutines.flow.toList
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.net.URI
+import java.util.concurrent.TimeUnit
+
+class RouteRefreshOnDemandTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    @get:Rule
+    val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
+
+    private lateinit var failedRefreshHandlerWrapper: FailByRequestMockRequestHandler
+    private lateinit var refreshHandler: MockDirectionsRefreshHandler
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private val twoCoordinates = listOf(
+        Point.fromLngLat(-121.496066, 38.577764),
+        Point.fromLngLat(-121.480279, 38.57674)
+    )
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        latitude = twoCoordinates[0].latitude()
+        longitude = twoCoordinates[0].longitude()
+        bearing = 190f
+    }
+
+    @Before
+    fun setup() {
+        setupMockRequestHandlers(
+            twoCoordinates,
+            R.raw.route_response_route_refresh,
+            R.raw.route_response_route_refresh_annotations,
+            "route_response_route_refresh"
+        )
+    }
+
+    @After
+    fun tearDown() {
+        failedRefreshHandlerWrapper.failResponse = false
+    }
+
+    @Test(timeout = 10_000)
```

Comment:
`sdkTest` fails with timeout of 30 seconds by default and it can be configured. TBH I'm not sure which way of configuring timeout is better 🙂